### PR TITLE
TELCODOCS-1007 Create VLANs based on container namespaced interfaces in OpenShift pods 

### PR DESCRIPTION
**Pull request by kquinn1204**
[TELCODOCS-1007]: [CNF-6185](https://issues.redhat.com//browse/CNF-6185) Apply multiple level chained CNI plugins

Version(s):
4.14 and main

Issue:
https://issues.redhat.com/browse/TELCODOCS-1007

Link to docs preview:

Preview: https://file.emea.redhat.com/kquinn/TELCODOCS-1007/networking/multiple_networks/configuring-additional-network.html#nw-about-configuring-master-interface-container_configuring-additional-network

QE review:
- [x] QE has approved this change.

Additional information:

### CODE REVIEW

File: modules/nw-chaining-multiple-interface-container.adoc [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+          "master": "ens5", <1>
+          "mtu": 1500,
+          "vlanId": 100,
+          "linkInContainer": true, <2>
+          "ipam": {"type": "whereabouts", "ipRanges": [{"range": "1.1.1.0/24"}]}
+        }
+      ]
+    }
+----
+
```
Comment (aireilly) [inline modules/nw-chaining-multiple-interface-container.adoc:117]:
```suggestion
```

Comment (aireilly) [reply modules/nw-chaining-multiple-interface-container.adoc]: remove the line altogether.

**Comment by kquinn1204** [comment]
Sorry @mlguerrero12 this is another one to review wrt to rootless DPDK etc.

**Comment by kquinn1204** [comment]
Hi @kononovn PTAL this PR falling under Rootless DPDK pods epic

File: modules/nw-chaining-multiple-interface-container.adoc [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx-pod
+  annotations:
+    k8s.v1.cni.cncf.io/networks: '[
+      {
+        "name": "sriov-network",
+        "namespace": "test-namespace",
+        "interface": "ens5" <1>
```
Comment (mlguerrero12) [inline modules/nw-chaining-multiple-interface-container.adoc:142]: this should be ext0

File: modules/nw-multus-ipvlan-object.adoc [hunk 3]
```diff
@@ -44,6 +46,10 @@ The following object describes the configuration parameters for the IPVLAN CNI p
 |`integer`
 |Optional: Set the maximum transmission unit (MTU) to the specified value. The default value is automatically set by the kernel.
 
+|`linkInContainer`
+|`boolean`
+|Specifies if the master interface is in the container network namespace or the main network namespace. To request the use of a container namespace master interface set the value to `true`
```
Comment (mlguerrero12) [inline modules/nw-multus-ipvlan-object.adoc:51]: Optional:

File: modules/nw-multus-macvlan-object.adoc [hunk 4]
```diff
@@ -39,6 +39,9 @@ The following object describes the configuration parameters for the macvlan CNI
 |`mtu`
 |`string`
 |Optional: The maximum transmission unit (MTU) to the specified value. The default value is automatically set by the kernel.
+|`linkInContainer`
+|`boolean`
+|Specifies if the master interface is in the container network namespace or the main network namespace.
```
Comment (mlguerrero12) [inline modules/nw-multus-macvlan-object.adoc:44]: same as above

File: networking/multiple_networks/configuring-additional-network.adoc [hunk 5]
```diff
@@ -28,7 +28,7 @@ For an additional network, IP addresses are provisioned through an IP Address Ma
 
 [NOTE]
 ====
-When deploying {product-title} nodes with multiple network interfaces on {rh-openstack-first} with OVN SDN, DNS configuration of the secondary interface might take precedence over the DNS configuration of the primary interface. In this case, remove the DNS nameservers for the subnet id that is attached to the secondary interface:
+When deploying {product-title} nodes with multiple network interfaces on {rh-openstack-first} with OVN SDN, DNS configuration of the secondary interface might take precedence over the DNS configuration of the primary interface. In this case, remove the DNS    for the subnet id that is attached to the secondary interface:
```
Comment (mlguerrero12) [inline networking/multiple_networks/configuring-additional-network.adoc:31]: why did you remove nameservers?

**Comment by mlguerrero12** [comment]
PR needs rebase

**Comment by kquinn1204** [comment]
I did rebase so should be good to go preview is https://file.emea.redhat.com/kquinn/TELCODOCS-1007/networking/multiple_networks/configuring-additional-network.html#nw-about-configuring-master-interface-container_configuring-additional-network

File: modules/nw-chaining-multiple-interface-container.adoc [hunk 6]
```diff
@@ -0,0 +1,173 @@
+// Module included in the following assemblies:
```
Comment (mlguerrero12) [inline modules/nw-chaining-multiple-interface-container.adoc:1]: I don't see the need to have a new section. It's clearer if you add to the previous section the content from line 14. I mean, just the example. No need to get into details of SRIOV

Comment (kquinn1204) [reply modules/nw-chaining-multiple-interface-container.adoc:1]: I merged into the previous section as per your suggestion.

**Comment by kquinn1204** [comment]
I will remove just was not sure you were happy with the preview as is in case it came back.

**Comment by kquinn1204** [comment]
Hi @mlguerrero12 I removed the file nw-chaining-multiple-interface-container.adoc from this PR and included the content after line 14 into the previous section see preview https://file.emea.redhat.com/kquinn/TELCODOCS-1007/networking/multiple_networks/configuring-additional-network.html#nw-about-configuring-master-interface-container_configuring-additional-network

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+// * networking/multiple_networks/configuring-additional-network.adoc
+
+:_module-type: CONCEPT
+
+[id="nw-about-configuring-master-interface-container_{context}"]
+= About configuring the master interface in the container network namespace
+
+In {product-title} 4.14 and greater the ability to allow users to create a MAC-VLAN, IP-VLAN and VLAN subinterface based on a master interface in a container namespace is now generally available.
+
+This feature allows you to create the master interfaces as part of the pod network configuration in a separate network attachment definition. YOU CAN THEN base the VLAN, MACVLAN or IPVLAN on this interface without requiring the knowledge of the network configuration of the node.
```
Comment (mlguerrero12) [inline modules/nw-about-configuring-master-interface-container.adoc:12]: YOU CAN THEN? Upperletters?

Comment (kquinn1204) [reply modules/nw-about-configuring-master-interface-container.adoc:12]: My bad gremlins on my keyboard.

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+
+The following worked example shows how to configure the set up illustrated in this diagram.
+
+.Creating VLANs
+image::345_OpenShift_config_additional_network_0823.png[Creating VLANs]
+
+.Prerequisites
+* SR-IOV Network Operator installed
+* Access to the cluster as a user with the `cluster-admin` role.
+* Install the OpenShift CLI (`oc`).
```
Comment (skopacz1) [inline modules/nw-about-configuring-master-interface-container.adoc:26]: If you accept my suggestion on line 24, I'd reword this one to be somewhat more parallel:

```suggestion
* The OpenShift CLI (`oc`) is installed.
```

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+To ensure the use of a container namespace master interface specify the  `linkInContainer` and set the value to `true` in the VLAN, MACVLAN or IPVLAN plugin configuration depending on the particular type of additional network.
+
+An example use case for utilizing this feature is to create multiple VLANs based on SR-IOV VFs. To do so, begin by creating an SR-IOV network and then define the network attachments for the VLAN interfaces.
+
+The following worked example shows how to configure the set up illustrated in this diagram.
+
+.Creating VLANs
+image::345_OpenShift_config_additional_network_0823.png[Creating VLANs]
+
+.Prerequisites
```
Comment (skopacz1) [inline modules/nw-about-configuring-master-interface-container.adoc:23]: Since all these prereqs seem to be sentence fragments, I would remove periods at the end of the last two items, [per the ISG](https://www.ibm.com/docs/en/ibm-style?topic=format-lists#punctuation-in-lists).

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+
+.Procedure
+
+. Create a dedicated container namespace where you want to deploy your pods by using the following command:
++
+[source,terminal]
+----
+$ oc new-project test-namespace
+----
+. Create a SR-IOV node policy.
```
Comment (skopacz1) [inline modules/nw-about-configuring-master-interface-container.adoc:36]: Just checking: is SR-IOV pronounced by spelling the letters out? If so, shouldn't the preceding article be "an"?

File: modules/nw-multus-ipvlan-object.adoc [hunk 11]
```diff
@@ -44,6 +46,10 @@ The following object describes the configuration parameters for the IPVLAN CNI p
 |`integer`
 |Optional: Set the maximum transmission unit (MTU) to the specified value. The default value is automatically set by the kernel.
 
+|`linkInContainer`
+|`boolean`
+|Optional: Specifies if the master interface is in the container network namespace or the main network namespace. To request the use of a container namespace master interface set the value to `true`.
```
Comment (skopacz1) [inline modules/nw-multus-ipvlan-object.adoc:51]: I'm a little confused by this parameter description. Is it set to true if the master interface is in one of those namespaces, and false if it's in neither? If that's the case, I'd suggest the following rewording:

```suggestion
|Optional: Specifies whether the master interface is in either the container network namespace or the main network namespace. To request the use of a container namespace master interface, set the value to `true`.
```

Or, is it set to true if the master interface is in the container network namespace and set to false if it's in the main network namespace? If this is the case, then I'd suggest the following:

```suggestion
|Optional: Specifies whether the master interface is in the container network namespace or in the main network namespace. To request the use of a container namespace master interface, set the value to `true`.
```

In both cases, I also replaced "if" with "whether", seeing as the ISG guidance (see "if versus whether" in the ["I" section of word usage](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#i)) suggests the use of "whether" in the case of discussing alternatives, and here it seems like you do need to choose between one of two options.

File: modules/nw-multus-macvlan-object.adoc [hunk 12]
```diff
@@ -40,6 +40,10 @@ The following object describes the configuration parameters for the macvlan CNI
 |`string`
 |Optional: The maximum transmission unit (MTU) to the specified value. The default value is automatically set by the kernel.
 
+|`linkInContainer`
+|`boolean`
+|Optional: Specifies if the master interface is in the container network namespace or the main network namespace.
```
Comment (skopacz1) [inline modules/nw-multus-macvlan-object.adoc:45]: Same question/suggestion as my comment on line 51 of modules/nw-multus-ipvlan-object.adoc

**Review by skopacz1 — COMMENTED** [review]
I left some comments, otherwise looks good to me! I am still new to reviewing, however, so please wait until @adellape approves of my review before implementing any feedback. Thanks!

File: 1
[the remark this answers is not in the document]
Comment (kquinn1204) [reply 1:1]: Weird that gremlins on my keyboard

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+An example use case for utilizing this feature is to create multiple VLANs based on SR-IOV VFs. To do so, begin by creating an SR-IOV network and then define the network attachments for the VLAN interfaces.
+
+The following worked example shows how to configure the set up illustrated in this diagram.
+
+.Creating VLANs
+image::345_OpenShift_config_additional_network_0823.png[Creating VLANs]
+
+.Prerequisites
+* SR-IOV Network Operator installed
```
[the remark this answers is not in the document]
Comment (kquinn1204) [reply modules/nw-about-configuring-master-interface-container.adoc:24]: Fixed in line with previous sections https://docs.openshift.com/container-platform/latest/networking/hardware_networks/configuring-sriov-device.html

File: modules/nw-about-configuring-master-interface-container.adoc (hunk 8 above)
Comment (kquinn1204) [reply modules/nw-about-configuring-master-interface-container.adoc:26]: Fixed in line with previous sections https://docs.openshift.com/container-platform/latest/networking/hardware_networks/configuring-sriov-device.html

File: modules/nw-about-configuring-master-interface-container.adoc (hunk 9 above)
Comment (kquinn1204) [reply modules/nw-about-configuring-master-interface-container.adoc:23]: Fixed in line with previous sections https://docs.openshift.com/container-platform/latest/networking/hardware_networks/configuring-sriov-device.html

File: modules/nw-multus-macvlan-object.adoc (hunk 12 above)
Comment (kquinn1204) [reply modules/nw-multus-macvlan-object.adoc]: Revised to make description clearer hopefully

File: modules/nw-multus-ipvlan-object.adoc (hunk 11 above)
Comment (kquinn1204) [reply modules/nw-multus-ipvlan-object.adoc]: revised

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+
+.Procedure
+
+. Create a dedicated container namespace where you want to deploy your pods by using the following command:
++
+[source,terminal]
+----
+$ oc new-project test-namespace
+----
+. Create an SR-IOV node policy.
```
Comment (adellape) [inline modules/nw-about-configuring-master-interface-container.adoc:36]: IBM SG wants colons for lead-in sentences going into substeps. I see some other instances of this in the PR so apply throughout.
```suggestion
. Create an SR-IOV node policy:
```

File: modules/nw-multus-ipvlan-object.adoc [hunk 15]
```diff
@@ -6,6 +6,8 @@
 // https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/getting-started-with-ipvlan_configuring-and-managing-networking#ipvlan-overview_getting-started-with-ipvlan
 :_content-type: REFERENCE
 
+:_module-type: CONCEPT
```
Comment (adellape) [inline modules/nw-multus-ipvlan-object.adoc:9]: Sorry, I'm not familiar with `_module-type`; is it required for TELCO docs? But either way, the value `CONCEPT` here seems at odds with the `_content-type: REFERENCE` above it.

**Review by adellape — COMMENTED** [review]
@skopacz1 @kquinn1204 Peer review comments LGTM, added a few more. 👍

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+kind: SriovNetworkNodePolicy
+metadata:
+ name: sriovnic
+ namespace: openshift-sriov-network-operator
+spec:
+ deviceType: netdevice
+ isRdma: false
+ needVhostNet: true
+ nicSelector:
+   vendor: "1d0f"
```
Comment (kononovn) [inline modules/nw-about-configuring-master-interface-container.adoc:52]: What is the vendor "1d0f"? I'm not sure if these vendor is supported
Can we use support mlx or intel vendor id instead: 15b3 or 8086?

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 17]
```diff
@@ -0,0 +1,175 @@
...
+----
+$ oc new-project test-namespace
+----
+. Create an SR-IOV node policy:
+
+.. Create an `SriovNetworkNodePolicy` object, and then save the YAML in the `sriov-node-network-policy.yaml` file:
++
+[source,yaml]
+----
+apiVersion: sriovnetwork.openshift.io/v1
```
Comment (kononovn) [inline modules/nw-about-configuring-master-interface-container.adoc:42]: Shoud we mentioned that this paritucal configuration works for mlx only because of:
 deviceType: netdevice
 isRdma: false

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+ nodeSelector:
+    feature.node.kubernetes.io/network-sriov.capable: "true"
+----
+
+.. Apply the YAML by running the following command:
++
+[source,terminal]
+----
+$ oc apply -f sriov-node-network-policy.yaml
+----
```
Comment (kononovn) [inline modules/nw-about-configuring-master-interface-container.adoc:67]: Should we add that it could take some time due to node reboot ?

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+
+. Create a pod definition by using the earlier specified networks and then save the YAML in the `pod-a.yaml` file. The pod definition must use the following networks annotation:
++
+[source,yaml]
+----
+apiVersion: v1
+kind: Pod
+metadata:
+  name: nginx-pod
```
Comment (kononovn) [inline modules/nw-about-configuring-master-interface-container.adoc:138]: Can we add namespace to pod annotation?
namespace: test-namespace

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+----
++
+<1> The name to be used as the master for the VLAN interface.
+
+. Apply the YAML by running the following command:
++
+[source,terminal]
+----
+$ oc apply -f pod-a.yaml
+----
```
Comment (kononovn) [inline modules/nw-about-configuring-master-interface-container.adoc:175]: Don't we want to show > "ip link show" inside container?
Please keep in mind that nginx-unprivileged doesn't have "ip" binary

Or may be we want to shouw output from:
>oc describe pods nginx-pod -n test-namespace
                },{
                    "name": "test-namespace/sriov-network",
                    "interface": "ext0",
                    "mac": "6e:a7:5e:3f:49:1b",
                    "dns": {},
                    "device-info": {
                        "type": "pci",
                        "version": "1.0.0",
                        "pci": {
                            "pci-address": "0000:d8:00.2"
                        }
                    }
                },{
                    "name": "test-namespace/vlan-100",
                    "interface": "ext0.100",
                    "ips": [
                        "1.1.1.1"
                    ],
                    "mac": "6e:a7:5e:3f:49:1b",
                    "dns": {}

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 21]
```diff
@@ -0,0 +1,175 @@
...
+        type: "RuntimeDefault"
+----
++
+<1> The name to be used as the master for the VLAN interface.
+
+. Apply the YAML by running the following command:
++
+[source,terminal]
+----
+$ oc apply -f pod-a.yaml
```
Comment (kononovn) [inline modules/nw-about-configuring-master-interface-container.adoc:174]: When I applied this manifest I've got following error:
`Error from server (Forbidden): error when creating pod-a.yaml": pods "nginx-pod" is forbidden: violates PodSecurity "restricted:v1.24": runAsNonRoot != true (pod must not set securityContext.runAsNonRoot=false), seccompProfile (pod or container "nginx-container" must set securityContext.seccompProfile.type to "RuntimeDefault" or "Localhost")`

I had to add following labels in order to make it work:
```
oc label namespace test-namespace pod-security.kubernetes.io/enforce=privileged --overwrite
oc label namespace test-namespace pod-security.kubernetes.io/audit=privileged
oc label namespace test-namespace pod-security.kubernetes.io/warn=privileged --overwrite
oc label namespace test-namespace security.openshift.io/scc.podSecurityLabelSync=false --overwrite 
```

File: modules/nw-about-configuring-master-interface-container.adoc (hunk 16 above)
Comment (kquinn1204) [reply modules/nw-about-configuring-master-interface-container.adoc:52]: I updated to specifically call out a mellanox device like here https://docs.openshift.com/container-platform/4.13/networking/hardware_networks/configuring-sriov-device.html

File: modules/nw-about-configuring-master-interface-container.adoc (hunk 17 above)
Comment (kquinn1204) [reply modules/nw-about-configuring-master-interface-container.adoc:42]: I added note up front.

**Comment by kquinn1204** [comment]
Hi @kononovn thanks for your feedback hopefully OK for an LGTM now.

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 22]
```diff
@@ -0,0 +1,258 @@
...
+[source,terminal]
+----
+$ oc apply -f vlan100-additional-network-configuration.yaml
+----
+
+. Create a pod definition by using the earlier specified networks and then save the YAML in the `pod-a.yaml` file. The pod definition must use the following networks annotation:
++
+[source,yaml]
+----
+apiVersion: v1
```
Comment (kononovn) [inline modules/nw-about-configuring-master-interface-container.adoc:148]: I failed to apply this resource due to:

```
Warning: resource namespaces/test-namespace is missing the kubectl.kubernetes.io/last-applied-configuration annotation which is required by oc apply. oc apply should only be used on resources created declaratively by either oc create --save-config or oc apply. The missing annotation will be patched automatically.
namespace/test-namespace configured
Error from server (Forbidden): error when creating "./pod-a.yaml": pods "nginx-pod" is forbidden: violates PodSecurity "restricted:v1.24": seccompProfile (pod or container "nginx-container" must set securityContext.seccompProfile.type to "Ru
```ntimeDefault" or "Localhost")

Comment (kononovn) [reply modules/nw-about-configuring-master-interface-container.adoc:148]: I suspect it's because of invalid namespace config. When I change it to:
```
---
apiVersion: v1
kind: Namespace
metadata:
  name: test-namespace
  labels:
    pod-security.kubernetes.io/enforce: privileged
    pod-security.kubernetes.io/audit: privileged
    pod-security.kubernetes.io/warn: privileged
    security.openshift.io/scc.podSecurityLabelSync: "false"
---
```
It works.

File: modules/nw-about-configuring-master-interface-container.adoc [hunk 23]
```diff
@@ -0,0 +1,258 @@
...
+<2> The `linkInContainer` parameter must be specified.
+
+. Apply the YAML by running the following command:
++
+[source,terminal]
+----
+$ oc apply -f vlan100-additional-network-configuration.yaml
+----
+
+. Create a pod definition by using the earlier specified networks and then save the YAML in the `pod-a.yaml` file. The pod definition must use the following networks annotation:
```
Comment (kononovn) [inline modules/nw-about-configuring-master-interface-container.adoc:144]: I think we need to add a note that we also add labels to namespace. Because manifest below contains 2 resources.
1-namespace with new label
2- pod manifest.

**Comment by kononovn** [comment]
lgtm
